### ISSUE TITLE
Highlight unexpected character encodings: "c" and "с"

### ISSUE THREAD

**Issue by lachlansimpson** [rstudio/rstudio#14485]
Hi Team, 

This request came from the support queue ticket 102957.

> Symbols "c" and "с" look identical, but actually belong to different encodings.
> If I use "с" in R, it'll cause a bug.
> Many IDE's highlight your code in a different color when you use a wrong symbol set, but not RStudio.
> Could you add this, please?

Sure enough:

```
$ cat c.txt 
c с
$ hexdump -C c.txt 
00000000  63 20 d1 81 20 0a                                 |c .. .|
00000006
$ hexdump -c c.txt 
0000000   c     321 201      \n                                        
0000006
```

Can we please add a highlight for unexpected character encodings?

Character 	с
Character name 	CYRILLIC SMALL LETTER ES
https://www.cogsci.ed.ac.uk/~richard/utf-8.cgi?input=1089&mode=decimal

**Comment by ronblum** [comment]
Agreed, these are practically indistinguishable to the untrained eye, but R and RStudio diagnostics know the difference:
```
> x <- c(1, 2, 3)
> x <- с(1, 2, 3)
Error in с(1, 2, 3) : could not find function "с"
```

![image](https://github.com/rstudio/rstudio/assets/31009952/d5922cc0-57d3-4099-8031-4671e8379d0b)

**Comment by Allexxann** [comment]
Also, several Greek letters also reuse English glyphs.
Meaning, the problem isn't endemic to Cyrillics.

**Comment by ronblum** [comment]
Can you provide a few examples? They're helpful for reproducing the issue. Thanks!

**Comment by lachlansimpson** [comment]
Thee Greek Question Mark issue is the one I thought of immediately. There are memes aplenty if you want to mine that part of the web, but you only need to see the glyph in question to understand why: `;`

https://www.compart.com/en/unicode/U+037E

**Comment by Allexxann** [comment]
I think I should provide an example of a different nature.
![изображение](https://github.com/rstudio/rstudio/assets/43035396/7501dc4d-ca61-412d-a150-5360cefe1f0f)


Here is my MATLAB IDE screenshot.
I wrote "LaTeX" alternating between Latin and Cyrillic letters (I don't have Greek installed) and some stuff without alternating, first unquoted, then quoted (I deliberately copied the same string).
The window automatically detects the encoding (keyboard layout? symbol set? sorry, I don't know the correct word) of the input symbols and colors them the same color as a warning.
Note that I haven't even hit "Enter" - the highlighting occurs immediately upon symbol input, both in console and code editor.
Also note that quotes cancel this behaviour - character vector can contain any values, that's their purpose.

This behaviour saved me from a lot of trouble, and I was surprised that this is not an accepted practice.

So, the correct solution would be not to search for homoglyphs and making exceptions for them, but simply color every letter not in "allowed" set, that set being "vanilla" Latin, aka the only letters making syntactycally correct names.
Math symbols are usually the same, but punctuation can indeed vary.

**Comment by ronblum** [comment]
My go-to is `РΡP` using Cyrillic, Greek, and Latin.

**Comment by Allexxann** [comment]
> My go-to is `РΡP` using Cyrillic, Greek, and Latin.

Good point.
